### PR TITLE
feat: allow empty CommitmentPrefix

### LINKED ISSUE #1006: FR: Allow empty CommitmentPrefix
## Feature Summary

In our use case, we have a Merkle trie which is dedicated for the exclusive use of the IBC module.  As a consequence, any trie path derived from IBC path starts at the root of the trie.  This means that in our case CommitmentPrefix is always empty.

## Proposal

Allow empty CommitmentPrefix values, i.e. remove `is_empty()` checks from CommitmentPrefix constructors, e.g. convert `TryFrom<Vec<u8>>` implementation. into `From<Vec<u8>>` implementation which just wraps the provided vector.

## Further discussion

After quick chat with @Farhad-Shabani a question arose whether verify_membership and verify_non_membership functions even need prefix argument. Should they be able to derive the prefix from the client state instead?

I’m not that familiar with the codebase to really answer that (and all I really care about is being able to have empty prefix as described above) but for some context when calling those two aforementioned methods the prefix arguments comes from:

* [Counterparty::prefix](https://docs.rs/ibc/latest/ibc/core/connection/types/struct.Counterparty.html#method.prefix) when getting prefix of the counterparty,
* [ValidationContext::commitment_prefix](https://docs.rs/ibc/latest/ibc/core/host/trait.ValidationContext.html#tymethod.commitment_prefix) when getting our prefix and
* Tendermint client further uses [ClientState::upgrade_path](https://docs.rs/ibc/latest/ibc/clients/tendermint/types/proto/v1/struct.ClientState.html#structfield.upgrade_path).

The last usage of the prefix could be eliminated in the sense that the trait function could not accept prefix and Tendermint would have its own internal method which takes prefix.

The case for the first two isn’t immediately obvious to me.  I guess it can be that each connection has its own prefix, but then I’m not sure how to reconcile it with `commitment_prefix` method not being dependent on connection.  Plus, if trie location depends on connection than I guess connection id should be included in the Path rather than using prefix for it.  Though I guess in principle it could be used for that.

It doesn’t help that the repository doesn’t seem to have any non-mock implementation of commitment_prefix (`git grep fn.\*\\bcommitment_prefix` revealed only declaration in trait and implementation for mock).

## Summary

To sum up, in our implementation prefix is always empty and I’d love if empty CommitmentPrefix would be allowed.

As for whether the prefix argument is needed to verify_membership and verify_non_membership, I can’t form well informed opinion.

### PR DESCRIPTION
**Pull request by Farhad-Shabani**
Closes: #1006
______

### PR author checklist:
- [X] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [X] Linked to GitHub issue.
- [X] Updated code comments and documentation (e.g., `docs/`).
- [X] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

**Comment by rnbguy** [comment]
It looks like the MSRV check failed on macOS runner. Farhad, can you please try on your machine?

```sh
cargo +1.71.1 build --all-features
```

**Comment by Farhad-Shabani** [comment]
> It looks like the MSRV check failed on macOS runner. Farhad, can you please try on your machine?

@rnbguy, Cosmwasm released a new version (v2.1.0) a few hours ago. The msrv creates a `cargo.lock` file with this version, pulling in the latest `derive_more-impl@1.0.0-beta.6`, which requires Rust 1.72.

I have locked Cosmwasm versions to allow only patch updates and left the upgrade to v2.1.0 for a separate PR.

File: ibc-core/ics03-connection/types/src/connection.rs [hunk 1]
```diff
@@ -392,8 +391,7 @@ impl TryFrom<RawCounterparty> for Counterparty {
                 .prefix
                 .ok_or(ConnectionError::MissingCounterparty)?
                 .key_prefix
-                .try_into()
-                .map_err(|_| ConnectionError::Client(ClientError::EmptyPrefix))?,
```
Comment (rnbguy) [inline ibc-core/ics03-connection/types/src/connection.rs:396]: Should we also remove the `ClientError::EmptyPrefix` variant?

**Review by rnbguy — APPROVED** [review]
LGTM :raised_hands:

File: ibc-core/ics03-connection/types/src/connection.rs (hunk 1 above)
Comment (Farhad-Shabani) [reply ibc-core/ics03-connection/types/src/connection.rs:396]: [df102af](https://github.com/cosmos/ibc-rs/pull/1274/commits/df102af5df6baa4ad76af1009e95e560e0554a70)

**Pull request merged**